src/core/model_config.proto

```diff
@@ -517,6 +517,57 @@ message ModelSequenceBatching
   repeated ControlInput control_input = 2;
 }
 
+//@@
+//@@.. cpp:var:: message ModelEnsembling
+//@@
+//@@   Model ensembling configuration. These settings specify the models that
+//@@   compose the ensemble and how data flows between the models.
+//@@
+message ModelEnsembling
+{
+  //@@  .. cpp:var:: message Step
+  //@@
+  //@@     Each step specifies a model included in the ensemble,
+  //@@     maps ensemble tensor names to the model input tensors,
+  //@@     and maps model output tensors to ensemble tensor names
+  //@@
+  message Step
+  {
+    //@@    .. cpp:var:: string model_name
+    //@@
+    //@@       The name of the model to execute for this step of the ensemble.
+    //@@
+    string model_name = 1;
+
+    //@@  .. cpp:var:: map<string,string> input_map
+    //@@
+    //@@     Map from ensemble tensor name to name of an input tensor on this
+    //@@     step's model. The ensemble tensor must have the same data type and
+    //@@     shape as the model input. Each model input must be assigned to
+    //@@     one ensemble tensor, but the same ensemble tensor can be assigned
+    //@@     to multiple model inputs.
+    //@@
+    map<string, string> input_map = 2;
+
+    //@@  .. cpp:var:: map<string,string> output_map
+    //@@
+    //@@     Map from ensemble tensor name to name of an output tensor on this
+    //@@     step's model. The data type and shape of the ensemble tensor will
+    //@@     be inferred from the model output. Each model output must be
+    //@@     assigned to one ensemble tensor, but the same ensemble tensor can
+    //@@     be assigned to multiple model outputs, where the content of
```

Comment:
We could allow one model output to be assigned to multiple ensemble tensors, but there is no reason to do that.. so for now we should disallow it. And there can be no overwriting, each ensemble tensor can appear in an output map only once. Once we have conditionals we will need to revisit this (though we will still likely want to keep the define-once invariant and instead use some other technique to handle it)

Reply:
For the overwriting, I was thinking about the same ensemble tensor are in different models' output maps. Then there will be conflict in the content of ensemble tensor if those models are run. But this can be avoided if we have conditionals, like making sure that only one of those model will be run.